### ISSUE TITLE
Theming improvements

### ISSUE THREAD

**Issue by kjk** [sumatrapdfreader/sumatrapdf#743]
@blake-mealey Some preliminary thoughts on theming.

1) Colors of tabs should be in sync with background/text colors, similar to how vscode works:

![vs-dark](https://cloud.githubusercontent.com/assets/7935/23735381/267e774e-043a-11e7-80ce-2495430e6f7c.PNG)
![vs-white](https://cloud.githubusercontent.com/assets/7935/23735382/26806202-043a-11e7-83b0-e0a8a217a3de.PNG)

2) Changing tab size leaves rendering artifacts that require resizing the window to fix. We should figure out how to fix that or not have variable tab size for now.

3) Bookmarks menu in dark theme doesn't look good:

![bookmarks-dark](https://cloud.githubusercontent.com/assets/7935/23735629/9e1f3008-043b-11e7-8614-721434834649.PNG)

Background color is good, text color is too dark.

In light theme, divider window is not visible:

![bookmarks-light-divider](https://cloud.githubusercontent.com/assets/7935/23735673/ec0e6324-043b-11e7-9a4a-6784f2734946.PNG)

4) I think that's the comment you made on the PR.

Changing colors of the ui (document.textColor, document.backgroundColor) should not be conflated with inverting colors of the PDF file (gRenderCache.textColor, gRenderCache.backgroundColor).

This behavior inherited from current code but to me those use cases are different.

Basically, a PDF has it's own colors. "Invert colors" options is a bit of a hack to re-color PDF bitmaps after it's been rendered. The intent was to provide "dark mode" given the fact that most PDFs are black text on white background.

I think this should be controlled independently from theming. There are two issues:

1. How to name it? "Invert colors" is not great. "Dark mode for PDF files" is a bit more explicit but still not great. Do other programs have similar functionality? If yes, how do they name it?

2. What colors should be used? Just go to white text/black background or have it conform to the ui theme?

Also related to that, it seems like we use gRenderCache.backgroundColor for UI where we should use theme colors.

5.) It would be nice to also adopt toolbar to theme colors. There already is support for changing background color in Toolbar.cpp (in handling WM_CTLCOLORSTATIC). We would also have to change text color and edit control colors.

**Comment by blake-mealey** [comment]
Good thoughts.

1. I was basing the "darker" theme off of Visual Studio 2015/2017, which is a bit different:

![](http://i.imgur.com/ed7IwFC.png)

However, I think it is a good idea to make the initial standard themes very simple and clean. If others want to make more interesting themes later, that's fine. Should I update the dark and light themes to look similar to those screenshots you posted then?

2. I noticed this, and I agree with keeping consistent tab sizes for now. I think I should make the default tab size slightly bigger for the dark/light themes described in (1) just because it will look better.

3. I hadn't even seen that menu, I'll look into it.

4. I just did some research and it looks like Adobe Reader and Foxit Reader both have this feature:

![](http://i.imgur.com/eVOTAkO.png)
![](http://i.imgur.com/w532QCo.png)

As you can see, they both put it under accessibility settings. Maybe "Contrast Mode" would be a good name for the feature. I suppose contrast mode should be accessible in any theme (even light), but the colours used could still be supplied by the current theme. This would just be to support continuity, since some themes would look weird with, for example, full black backgrounds on the documents. Also, I don't think white text on full black background is very nice to read compared with the colours in VS for example.

I am not quite sure where the gRenderCache comes into this (is it for the document thumbnails?), or how you want it to be handled.

5. This would be great! It would also be great (although not as necessary) to add theme colours to all of the menus. I have no idea how difficult that would be.

This also brings up the issue that toolbar icons would have to change colours. With the dark theme, toolbar icons would have to be white/light, and with the light theme, they would have to be black/dark. I think this would be best solved by implementing the toolbar that was discussed in issue https://github.com/sumatrapdfreader/sumatrapdf/issues/734 and using svg icons which we can then decide which colour to draw them with at render time.


I'm very happy to help with all of this, although I don't know how much free time I have coming up (I'm a full time university student and I've just had a light couple of weeks). I also have no clue where to start on (5) which seems like quite a large project, so I would need some guidance/help with it. I'll work on the other points as I have time though :)

**Comment by kjk** [comment]
> Should I update the dark and light themes to look similar to those screenshots you posted then?

Please do. Also, feel free to add yourself to AUTHORS file.

Make sure to submit PRs relative to the latest master. Given that you've modified your fork's master relative to Sumatra master, the simplest way to get that might be to re-fork.

> I am not quite sure where the gRenderCache comes into this (is it for the document thumbnails?), or how you want it to be handled.

gRenderCache is how we implement color swapping in PDF files: they get rendered by render cache and then we change colors in the bitmap based on gRenderCache.backgroundColor etc.

I just noted that in current code we use gRenderCache.backgroundColor in context where we should   be probably using theme colors (in Canvas.cpp).

Both Adobe and FoxIt have too much UI for this. For Sumatra maybe a simplified approach that combines menu item e.g. "Replace PDF document colors" that is on/off and settings for providing background replacement/text replacement colors maybe with option to use theme colors.

I'm not convinced there's value in "Use Windows Colors Scheme".

> I'm very happy to help with all of this, although I don't know how much free time I have coming up (I'm a full time university student and I've just had a light couple of weeks).

No pressure and I'm happy to help.

**Comment by VictorVG** [comment]
In to Git 48c2071c6727f bug #745 is exist, but partial - on some then switched theme to Dark or Darer see original color, or re-rendering have timeout then I see color error #745. I checking x86 and AMD64 build compile use VC++2015 Update 3 (VC++2017 have many problem's - new patch for msbuild $INSTDIR/2017/"vs_edition_name"/$MsBuildDir , bug in VS2017 rand_s() function realization, not have any Registry entries for configuration, e.t.c. ... ).

**Comment by VictorVG** [comment]
v3.2 Git fceb4e59b6 bug #745 is exist, but it's interesting - if you use the settings

```
FixedPageUI [
SelectionColor = # f5fc0c
WindowMargin = 2 4 2 4
PageSpacing = 4 4
TextColor = # 000000
BackgroundColor = #ffffff
]
EbookUI [
FontName = Georgia
FontSize = 12.5
UseFixedPageUI = true
TextColor = # 5f4b32
BackgroundColor = # fbf0d9
]
ComicBookUI [
WindowMargin = 0 0 0 0
PageSpacing = 4 4
CbxMangaMode = false
]
ChmUI [
UseFixedPageUI = true
]
```
Then on the Dark / Darker themes the color errors appear chaotically - in one tab there can be shown in others not, if for the EbookUI you put the background xffffff - the reproducibility of the error is 100%, even where it did not seem to exist.

It looks funny, but the bug-reports we sprinkle. :)

**Comment by jg-github** [comment]
Sorry to be boring, but is there any way to get the nice golden yellow background back, please?

Thanks.

**Comment by VictorVG** [comment]
Observations on the assembly v3.2 Git 2de1381a00 Win32 / Win64

By observing the behavior of the * FixedPageUI = {true | false} option, I noticed that the manifestation of the error "Overlaying the background as a transparent texture on the image" and then distorting its colors for Dark and Darker themes is 100% when this setting is set to true for any From the rendering engines, if the option EbookUI :: UseFixedPageUI = true, switching the themes Light -> Dark or Light -> Darker results in an error on the sdjvuspec.djvu file (it may be incorrectly recognized by the engine) in the form of color distortion, but the reverse switching of the themes This error eliminates the error.

If you set EbookUI :: UseFixedPageUI = false, then the probability of color distortion and subsequent rendering of documents with the output of the colors specified in it tends to zero.
![dark](https://cloud.githubusercontent.com/assets/906886/24080891/09fe14ee-0cb9-11e7-9a0e-2a35155ab93c.png)
![darker](https://cloud.githubusercontent.com/assets/906886/24080892/0a007d56-0cb9-11e7-9ac5-371607987a43.png)
![light](https://cloud.githubusercontent.com/assets/906886/24080894/0a02cdae-0cb9-11e7-89f8-bbf516a926ed.png)
![light_original](https://cloud.githubusercontent.com/assets/906886/24080893/0a0253a6-0cb9-11e7-9c05-59366f8211d3.png)

Test kit included executable files [**Test/7z**](https://www.upload.ee/files/6802114/Test.7z.html) ...

**Comment by JJenkx** [comment]
In Windows, my setup is all black boarders and black start-up screen and total black in full-screen with white text.

Setting/Advanced Options, add line **GradientColors** to **FixedPageUI** then choose your hex color. Here is mine setup for all black and white.

```
MainWindowBackground = #000000
EscToExit = false
ReuseInstance = false
UseSysColors = false
RestoreSession = true

FixedPageUI [
	TextColor = #ffffff
	BackgroundColor = #000000
	SelectionColor = #f5fc0c
	WindowMargin = 2 4 2 4
	PageSpacing = 4 4
	GradientColors = #000000
]
```

https://i.stack.imgur.com/vMHOQ.png

**Comment by blake-mealey** [comment]
@kjk Hello! I have been completely inactive for many months now and I apologize for that. I see that you have (understandably) removed my changes for now since they were very incomplete. I was wondering if they were put somewhere accessible? I am thinking of working on my own fork/branch to get it more stable/correct now that I have some more free time.

**Comment by kjk** [comment]
@blake-mealey The code is still there (in Theme.h and Theme.cpp), just not enabled (ENABLE_THEME is not defined).

One of the issues is that the set of colors used is not well defined so I did https://github.com/sumatrapdfreader/sumatrapdf/pull/793/files which centralizes colors in Color.h/Color.cpp.

It's written so that GetAppColor() can use colors from themes.

**Comment by blake-mealey** [comment]
Ahh, I see. Now if I was to add themes support to Color.cpp, how would you like that to be done? At what level does it override other options? Should it use the same system as we were before (ex: GetCurrentTheme()->document.canvasColor)?

I think we should move this discussion to somewhere more appropriate, but I would like to really flesh out the requirements for themes in Sumatra. Previously I was just doing my own thing and not really taking in the big picture or thinking about how existing features should be integrated, etc.

**Comment by kjk** [comment]
The sketch of how to reach into themes is https://github.com/sumatrapdfreader/sumatrapdf/blob/master/src/Colors.cpp#L254

To me the hard part is to identify and name the colors. I don't think Colors.hpp covers all the colors and I don't think the colors are precisely defined.

**Comment by andrewrproper** [comment]
Hi, I haven't contributed to this project, but I'm thinking about how I could contribute a very small change that might allow the user to configure multiple themes.

I currently use the Advanced Settings to configure a theme, but I would like to be able to easily switch between one of three themes: Dark, Light, Sepia.

I'm thinking that we could make 2 changes to Advanced Settings:

1. add a list of Themes with colours, fonts etc
2. add a key which specifies which theme is active
  - a more ambitious idea would be to add a drop-down to the Options menu which allows picking one of the themes listed in Advanced Settings. 


for example:

```
SelectedTheme: "Dark"
Themes [
   Dark [
      FixedPageUI [
         TextColor = #ffffff
         BackgroundColor = #202020
         SelectionColor = #f5fc0c
         WindowMargin = 2 4 2 4
         PageSpacing = 4 4
      ]
      EbookUI [
         FontName = Tahoma
         FontSize = 13.5
         TextColor = #e0e0e0
         BackgroundColor = #202020
         UseFixedPageUI = false
      ]
   ]
   Sepia [
      FixedPageUI [
         TextColor = #000000
         BackgroundColor = #fbf0d9
         SelectionColor = #f5fc0c
         WindowMargin = 2 4 2 4
         PageSpacing = 4 4
      ]
      EbookUI [
         FontName = Tahoma
         FontSize = 13.5
         TextColor = #5f4b32
         BackgroundColor = #fbf0d9
         UseFixedPageUI = false
      ]
   ]
   Light [
      FixedPageUI [
         TextColor = #000000
         BackgroundColor = #ffffff
         SelectionColor = #f5fc0c
         WindowMargin = 2 4 2 4
         PageSpacing = 4 4
      ]
      EbookUI [
         FontName = Tahoma
         FontSize = 13.5
         TextColor = #5f4b32
         BackgroundColor = #fbf0d9
         UseFixedPageUI = false
      ]
   ]
]
```

**Comment by andrewrproper** [comment]
I also want to note that my idea is about **document display themes** only, not about application theming. I don't mind the existing application theming, and it works well to use the existing advanced settings to apply one of my themes to a document. 

The annoying part, which I want to solve, is that switching between my three themes requires editing the advanced settings file and pasting the relevant sections in from other text files that I use to keep my theme settings.

**Comment by andrewrproper** [comment]
Here's an example screenshot of the "Dark" theme I'm currently using:

![image](https://user-images.githubusercontent.com/8821850/105398106-f0c9cb00-5bef-11eb-8e68-0d7a72634dce.png)

**Comment by GitHubRulesOK** [comment]
Poking a set of alternative values responds instantly and is fairly easy using a text substitution tool.
I had experimented by most simply prepending to settings but that is effectively blocked by the use of a BOM,
The alternative example solution I gave in forum/here in github is to use a "find and replace" tool with the existing setting, that's more complex with 3 choices rather than 2 but still do able.

 The simplest solution is to just use the command line setting with 3 desktop shortcuts so no programming but would need to exit and re-enter with the alternative paper theme.
First a shortcut to yellow paper then close and reopen as per your sepia setting 
`SumatraPDF.exe -set-color-range #000000 #fbf0d9`
![image](https://user-images.githubusercontent.com/12754913/105410800-66528d00-5c2a-11eb-9017-f97663f013a2.png)

**Comment by andrewrproper** [comment]
@GitHubRulesOK  I can see that working as a feasible workaround, thanks. But I was hoping to make something that is

1. built-in and therefore available for all users to use, in whatever version they have (portable or installed)
2. simple/intuitive to use

**Comment by andrewrproper** [comment]
@GitHubRulesOK now that you mentioned the CLI options, I see this one might be usable to have multiple settings files:

```
-appdata <directory> : set custom directory where we'll store SumatraPDF-settings.txt file and thumbnail cache
```

**Comment by andrewrproper** [comment]
I think my ideal way of solving this would be to have a theme selector right on the main menu bar, which can be toggled per-document, and which opens the next new document in the currently displayed document's theme.

That might be more complex to implement.

But would that go against any design/UI principles that you have for this project so far?

**Comment by GitHubRulesOK** [comment]
-appdate uses different historys and although I have suggested it may be used for seperate day/night settings is not as simple or as useful as simply starting with desired colours. The theme selector in menu is what was used here (without user settings) but eventually dropped because there were too many different effects that introduced UI artifacts. What is really needed for theming is an OS that is not the 10th incarnation 🙁

**Comment by andrewrproper** [comment]
@GitHubRulesOK thanks for helping me understand. Yes I can see that separate variables would likely be needed, as a first step. I didn't know that they are shared. I was hoping to have themes that affect only the document display. And to me, as a user, that's how it seems to act when I change advanced settings under windows 10. It updates the document colors, but the ui stays the same.

**Comment by andrewrproper** [comment]
@GitHubRulesOK  and others, I had another idea of how this could be done.

After looking at the code, I saw that the build process uses Go to generate the C++ code that defines the structs that define the data structures for the Settings txt file. This seems complex, but I imagine there is a good reason for this setup.

I could edit these data structures, and add a way to have multiple themes defined in the settings file, with one selected (by default, theme "Default").

OR I could create a separate script/app that would read a set of external theme settings, and update the settings txt file externally. It would be independant from the main app. 

There are pros and cons to this idea, of course. 

pro:
- less change to the main app
- 2nd app doesn't have to be tightly coupled to main app, and could be easier to extend/develop because of this

con:
- probably harder to maintain, because changes to the main app's settings structure will break the 2nd app


It's a similar idea to the GTK theme chooser I've seen in Linux environments.

**Comment by GitHubRulesOK** [comment]
There have been many ways to simply change the settings, and since it does not involve disturbing any full attempt at wading through all the colour variables, then I favour the simplest idea of changing settings.txt on the fly, especially as it responds instantly. It is more complex when setting both Ebook and FixedPage since some users will want both the same, whilst others may want a difference. What most users expect is the ability to change surround (best done with Windows Blinds) but failing that then as mentioned when changing "Back to Black" they don't want the Page/Find to be white on light. Again your average GUI user expects an elaborate integrated Colour Picker for the one time a year they want a change of theme.
So if you can split out the Find text color from the page text colour that would be a significant step. As to providing a colour picker that could be left to last, what then falls between the two is a reliable way to (allowing for line numbers changing) find and replace the two TextColor and background colour entries with same or different values perhaps from an ini file of presets (most of that is possible using a scripting approach such as AutoHotKey, but any open source script would do) My problems getting a simple command line to work stem from the need to ensure the correct settings.txt lines are overwritten without corrupting the overall settings file thus find #000000 and replace with #ffffff needs to be robust enough to handle any bad or out of order entries added by global users but SumatraPDf will usually reorder FixedPageUI [...] to come before EbookUI [...] so the two TextColors should also follow

**Comment by andrewrproper** [comment]
Yes, you bring up good points,  @GitHubRulesOK . Do you know if @kjk is around to comment on what might be the preferred approach?

I can make a program that will read in the file, make the needed changes, send atomically over write the file. But the can still be race conditions, with the main app editing the file and the theme app trying to do so at the same time. 

I should probably think this over and also do more reading of the code, before choosing a plan. Brainstorming here helps. 

I agree with you that we need to:

- [ ] separate the variables, so that pdf background color is separate from app window background color, etc. 

I might create a separate issue here for that, and link it here. 

That would open up some flexibility dippe making further document theme related changes. 

Then if people want to make app theme changes, that can be done as well, and does not need to be tied to document theme. 

Thanks

**Comment by GitHubRulesOK** [comment]
SumatraPDF tolerates edits to the settings.txt, and I must admit I have never tested how much external edits will be overwritten whilst it is running, My past approach to this has usually (but not always) been based on using SumatraPDF ExternalViewer command to initiate the change thus execution starts from within the same process which may help.
See the comments in https://github.com/sumatrapdfreader/sumatrapdf/issues/1027 then follow the link to https://forum.sumatrapdfreader.org/t/light-dark-theme-day-night-mode/424/10?u=githubrulesok (sorry the screenshots are missing) but note using that method of toggeling, the external call required a work around triggering settings changes. There is no need to use two externalviewers slots if your approach offers three or more choices for the price of one.

**Comment by GitHubRulesOK** [comment]
P.S forgot to mention/remind you apart from the complaint it can not correct inversion of pictures (expect many complaints on that score) there is also the question of spotting GradientColors in SOME cases and they can be one two or three values.

**Comment by kjk** [comment]
The issue with theming is much more complicated than changing 2 colors.

See (currently disabled) Theme.h and Theme.cpp. There are many colors that need to be defined and set for a given theme.

The reason this code is currently disabled is because there are even more colors that need to be defined and the code modified to use those colors. Toolbar, tree view, menus, tabs, each document type etc.

Theme.h/.cpp was an effort to define those colors but it wasn't finished by the person who started this effort and it's not a priority for me.

So that would have to be finished before we can talk about changing those colors via settings file.

**Comment by andrewrproper** [comment]
Thanks for the reply, @kjk . This is of course discouraging because I didn't want to create the theme for the whole application. I only wanted to create themes for the document display, which I expect should be much simpler if kept separate from the rest of the application.

**Comment by rwperrott** [comment]
This should eventually deprecate #79, because it's a better solution.
